sdk/nodejs/provider/experimental/analyzer.ts

```diff
@@ -584,6 +588,26 @@ function isInput(type: typescript.Type): boolean {
     return hasOutput && hasPromise && hasOther;
 }
 
+function isResourceReference(type: typescript.Type, checker: typescript.TypeChecker): boolean {
+    if (type.isClassOrInterface()) {
+        const baseTypes = checker.getBaseTypes(type as typescript.InterfaceType);
+        for (const baseType of baseTypes) {
+            const symbol = baseType.getSymbol();
+            const matchesName =
+                symbol?.escapedName === "CustomResource" ||
+                symbol?.escapedName === "ComponentResource" ||
+                symbol?.escapedName === "Resource";
+            const sourceFile = symbol?.declarations?.[0].getSourceFile();
+            const matchesSourceFile =
+                sourceFile?.fileName.endsWith("resource.ts") || sourceFile?.fileName.endsWith("resource.d.ts");
```

Comment:
Could you please explain why this check is needed?

Reply:
Yes, it's a pattern we've repeated throughout her.  We don't have the actual class here, so we can't check if this is a `pulumi.Resource`, but instead we just have the name.  So we could in theory match a `Resource` class from the users program.  To make that less likely, we also check that the filename matches where pulumi defines the `Resource` class.  Unfortunately this is the best way we found to make sure of this so far.